libs/pdi-scanner/src/rust/scanner.rs

```diff
@@ -87,138 +81,122 @@ impl Scanner {
 
-        let mut transfer_pool = rusb_async::TransferPool::new(self.device_handle.clone());
+        let mut in_primary_queue = self.scanner_interface.bulk_in_queue(ENDPOINT_IN_PRIMARY);
+        let mut in_image_data_queue = self.scanner_interface.bulk_in_queue(ENDPOINT_IN_IMAGE_DATA);
+        let device_handle = self.scanner_interface.clone();
         let default_timeout = self.default_timeout;
 
-        self.thread_handle = Some(thread::spawn({
-            move || {
-                tracing::debug!("Scanner thread started; submitting initial IN endpoint transfers");
-                transfer_pool
-                    .submit_bulk(ENDPOINT_IN_PRIMARY, vec![0; BUFFER_SIZE])
-                    .unwrap();
+        tokio::spawn(async move {
+            tracing::debug!("Scanner thread started; submitting initial IN endpoint transfers");
+
+            in_primary_queue.submit(RequestBuffer::new(DEFAULT_BUFFER_SIZE));
 
-                transfer_pool
-                    .submit_bulk(ENDPOINT_IN_IMAGE_DATA, vec![0; BUFFER_SIZE])
-                    .unwrap();
+            // Submit two initial transfers to the image data transfer queue so
+            // that we always have one ready to go while we process the other.
+            // We want image data transfer to be as fast as possible so the
+            // scanner doesn't overflow its internal buffer.
+            in_image_data_queue.submit(RequestBuffer::new(IMAGE_BUFFER_SIZE));
+            in_image_data_queue.submit(RequestBuffer::new(IMAGE_BUFFER_SIZE));
 
-                loop {
-                    if let Ok(()) = stop_rx.try_recv() {
+            loop {
+                tokio::select! {
+                    biased;
+
+                    _ = &mut stop_rx => {
                         tracing::debug!("Scanner thread received stop signal");
                         break;
                     }
 
-                    match host_to_scanner_rx.try_recv() {
-                        Ok((id, packet)) => {
-                            let bytes = packet.to_bytes();
-                            tracing::debug!(
-                                "sending packet: {packet:?} (data: {data:?})",
-                                data = String::from_utf8_lossy(&bytes)
-                            );
-
-                            transfer_pool.submit_bulk(ENDPOINT_OUT, bytes).unwrap();
-
-                            // wait for submit_bulk to complete
-                            transfer_pool
-                                .poll_endpoint(ENDPOINT_OUT, default_timeout)
-                                .unwrap();
-
-                            host_to_scanner_ack_tx.send(id).unwrap();
-                        }
-                        Err(mpsc::TryRecvError::Empty) => {}
-                        Err(e) => {
-                            tracing::error!("Error receiving outgoing packet: {e}");
+                    completion = in_image_data_queue.next_complete() => {
```

Comment:
Note that I reordered the future match patterns so that we check for image data first, then incoming events, then outgoing messages. I figured that might be a little bit safer (e.g. if the scanner rapidly sends the EndScan event after sending the last image data packet).